blob/src/main/java/io/crate/blob/v2/BlobIndex.java

```diff
@@ -65,24 +71,73 @@ public static String stripPrefix(String indexName) {
 
     private final Map<Integer, BlobShard> shards = new ConcurrentHashMap<>();
     private final Path globalBlobPath;
+    private final ESLogger logger;
 
-
-    BlobIndex(@Nullable Path globalBlobPath) {
+    BlobIndex(ESLogger logger, @Nullable Path globalBlobPath) {
         this.globalBlobPath = globalBlobPath;
+        this.logger = logger;
     }
 
     void createShard(IndexShard indexShard) {
         shards.put(indexShard.shardId().id(), new BlobShard(indexShard, globalBlobPath));
     }
 
     BlobShard removeShard(ShardId shardId) {
+        Path blobRoot = null;
+        if (shards.size() == 1) {
+            blobRoot = this.retrieveBlobRootDir(shardId);
+        }
         BlobShard shard = shards.remove(shardId.id());
         if (shard != null) {
             shard.deleteShard();
         }
+        if (blobRoot != null) {
+            this.deleteIndex(blobRoot, shardId.getIndex());
+        }
+
         return shard;
     }
 
+    /**
+     * Traverse the path down until the shard index was found
+     * and return the root path of the blob directory.
+     */
+    private Path retrieveBlobRootDir(ShardId shardId) {
+        Path blobDir = this.getShard(shardId.getId()).getBlobDir();
+        Path resultDir = Paths.get(blobDir.toString());
+        for (Iterator i = blobDir.iterator(); i.hasNext();) {
+            String resultDirStr = resultDir.toString();
+            int pos = resultDirStr.lastIndexOf("/");
+            String currentElement = resultDirStr.substring(pos + 1, resultDirStr.length());
+            if (currentElement.equals(shardId.getIndex())) {
+                break;
+            }
+            resultDir = resultDir.getParent();
```

Comment:
Wouldn't it be possible to just call getParent until `endsWith(indexName)` and get rid of the manual path split logic above?